_posts/2015-11-23-elm-for-javascript-developers.md

```diff
@@ -0,0 +1,111 @@
+- The function has no side effects, and does not mutate anything or change the state of the world around it.
+
+It's of course possible to create functions like this in JavaScript, and you can make it a rule in your application that functions should be pure. Elm enforces it due to its immutable nature, and this means it's impossible for impure functions to sneak into your code base, either through code you write or through code in a 3rd party library you're using.
+
+## Types
+
+Elm is a statically typed language. This might sound off putting, but it actually leads to far more robust applications. In Elm every value has a type.
+
+```elm
+"Hello World" - String Type
+True - Boolean type
+3 - number type
+3.14 - Float type
+[1, 2, 3] - List number type
+```
+
+You might think this is similar to JavaScript, and you'd be right. In JavaScript (and every other programming language), values have a particular type. The crucial difference comes when we pair this type system with functions. In JavaScript you might have a function that can take multiple types, and return multiple types:
+
+```js
+someMadeUpFn('Foo') => 5
+someMadeUpFn(5) => 'Foo'
+someMadeUpFn({ name: 'Jack' }) => { name: 'jack' }
+```
+
+There's no restriction on the types of the arguments that `someMadeUpFn` takes, and there's no restrictions on the type of the value it returns either. In Elm we have to explicitly declare all the types (actually, we could leave it up to the compiler to infer the types, but it's best practicse to declare them). The below code creates a function `square` that takes an integer and returns another.
+
+```elm
+square : Int -> Int
+square x = x * x
+```
+
+If I were to write the same function in JavaScript, I'd write:
+
+```js
+function square(x) {
+  return x * x;
+}
+```
+
+Notice the first line of our Elm function:
+
+```elm
+square : Int -> Int
+```
+
+This is a __type annotation__ that tells Elm that this function will take one argument which will be an integer, and return a value that's also an integer. That means if we try to call this function with a different data type, we'll get an error. Although this restriction can take some time to adjust to, it actually leads to much cleaner code that's easier to work with and follow. It also means you realise straight away if you're using a function incorrectly.
+
+## Compiling
+
+Above we discussed how if you try to call a function with the wrong type, you'll get an error. It's better than that though, because you'll get these errors at __compile time__. Elm as a language compiles to JavaScript, and we need to run the compiler to generate JavaScript from our Elm code. Elm's compiler is smart, and is able to check the types of values when it compiles our code into JavaScript. For example, if I take this Elm code and try to compile it, we'll get an error. Don't worry about the specifics of the syntax, but know that this code will call the `square` function with the argument `"Hello"`.
```

Comment:
`Above we noted that trying to call a function with the wrong types causes an error.`